src/haxelib/api/Installer.hx

```diff
@@ -0,0 +1,852 @@
+		?log:Null<(msg:String, priority:LogPriority)-> Void>,
+		?confirm:Null<(msg:String)->Bool>,
+		?logDownloadProgress:Null<(filename:String, finished:Bool, cur:Int, max:Null<Int>, downloaded:Int, time:Float) -> Void>,
+		?logInstallationProgress:Null<(msg:String, current:Int, total:Int) -> Void>
+	) {
+		_log = log;
+		_confirm = confirm != null ? confirm : (_)-> {true;};
+		_logDownloadProgress = logDownloadProgress;
+		_logInstallationProgress = logInstallationProgress;
+	}
+
+	public inline function log(msg:String, priority:LogPriority = Default):Void {
+		if (_log != null)
+			_log(msg, priority);
+	}
+
+	public inline function confirm(msg:String):Bool {
+		return _confirm(msg);
+	}
+
+	public inline function logInstallationProgress(msg:String, current:Int, total:Int):Void {
+		if (_logInstallationProgress != null)
+			_logInstallationProgress(msg, current, total);
+	}
+
+	public inline function logDownloadProgress(filename:String, finished:Bool, cur:Int, max:Null<Int>, downloaded:Int, time:Float):Void {
+		if (_logDownloadProgress != null)
+			_logDownloadProgress(filename, finished, cur, max, downloaded, time);
+	}
+
+	public inline function getDownloadProgressFunction():Null<(filename:String, finished:Bool, cur:Int, max:Null<Int>, downloaded:Int, time:Float) -> Void> {
+		return _logDownloadProgress;
+	}
+}
+
+/**
+	Like `LibFlagData`, but `None` is not an option.
+
+	Always contains enough information to reproduce
+	the same library installation.
+**/
+enum InstallData {
+	Haxelib(version:SemVer);
+	VcsInstall(version:VcsID, vcsData:VcsData);
+}
+
+private class AllInstallData {
+	public final name:ProjectName;
+	public final version:Version;
+	public final isLatest:Bool;
+	public final installData:InstallData;
+
+	function new(name:ProjectName, version:Version, installData:InstallData, isLatest:Bool) {
+		this.name = name;
+		this.version = version;
+		this.installData = installData;
+		this.isLatest = isLatest;
+	}
+
+	public static function create(name:ProjectName, libFlagData:LibFlagData, versionData:Null<Array<SemVer>>):AllInstallData {
+		if (versionData != null && versionData.length == 0)
+			throw new InstallationException('The library $name has not yet released a version');
+
+		return switch libFlagData {
+			case None:
+				final semVer = getLatest(versionData);
+				new AllInstallData(name, semVer, Haxelib(semVer), true);
+			case Haxelib(version) if (!versionData.contains(version)):
+				throw new InstallationException('No such version $version for library $name');
+			case Haxelib(version):
+				new AllInstallData(name, version, Haxelib(version), version == getLatest(versionData));
+			case VcsInstall(version, vcsData):
+				new AllInstallData(name, version, VcsInstall(version, vcsData), false);
+		}
+	}
+}
+
+private class AlreadyUpToDate extends InstallationException {}
+
+private function getLatest(versions:Array<SemVer>):SemVer {
+	if (versions.length == 0)
+		throw 'Library has not yet released a version';
```

Comment:
This is probably caught in the Main.hx. Perhaps it should be turned into a proper exception type because it is what people using the Installer class will see if a library has no releases.

Reply:
Yeah, this message actually appears two other times as `throw new InstallationException('The library $library has not yet released a version');`.